src/lib/run-dev.js

```diff
@@ -0,0 +1,165 @@
+/*
+Copyright 2019 Adobe. All rights reserved.
+This file is licensed to you under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License. You may obtain a copy
+of the License at http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software distributed under
+the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR REPRESENTATIONS
+OF ANY KIND, either express or implied. See the License for the specific language
+governing permissions and limitations under the License.
+*/
+/* eslint-disable no-template-curly-in-string */
+const aioLogger = require('@adobe/aio-lib-core-logging')('@adobe/aio-cli-plugin-app:runDev', { provider: 'debug' })
+const rtLib = require('@adobe/aio-lib-runtime')
+const rtLibUtils = rtLib.utils
+const vscode = require('./vscode')
+const bundle = require('./bundle')
+const serve = require('./serve')
+const Cleanup = require('./cleanup')
+const runDevLocal = require('./run-dev-local')
+
+const buildActions = require('./build-actions')
+const deployActions = require('./deploy-actions')
+const actionsWatcher = require('./actions-watcher')
+
+const utils = require('./app-helper')
+const execa = require('execa')
+const { run: logPoller } = require('./log-poller')
+
+/** @private */
+async function runDev (args = [], config, options = {}, log = () => {}) {
+  // note: args are considered perfect here because this function is only ever called by the `app run` command
+
+  /* parcel bundle options */
+  const bundleOptions = options.parcel || {}
+  /* skip actions */
+  const skipActions = !!options.skipActions
+  /* fetch logs for actions option */
+  const fetchLogs = options.fetchLogs || false
+
+  // control variables
+  const hasFrontend = config.app.hasFrontend
+  const withBackend = config.app.hasBackend && !skipActions
+  const isLocal = !options.devRemote // applies only for backend
+
+  aioLogger.debug(`hasFrontend ${hasFrontend}`)
+  aioLogger.debug(`withBackend ${withBackend}`)
+  aioLogger.debug(`isLocal ${isLocal}`)
+
+  let frontEndUrl
+
+  // state
+  let devConfig = config // config will be different if local or remote
+  devConfig.envFile = '.env'
+
+  const cleanup = new Cleanup()
+
+  // bind cleanup function
+  process.on('SIGINT', async () => {
+    // in case app-scripts are eventually turned into a lib:
+    // - don't exit the process, just make sure we get out of waiting
+    // - unregister sigint and return properly (e.g. not waiting on stdin.resume anymore)
+    try {
+      await cleanup.run()
+      log('exiting!')
+      process.exit(0) // eslint-disable-line no-process-exit
```

Comment:
do we need to do a hard exit?  Can't we just set exist code and let normal process-exit continue?
I think simply setting process.exitCode is sufficient here.

Reply:
Normally it would be if all all threads are cleaned up properly and not hanging on to resources. I can't find the straggler here, so we'll need to keep this in for now.

Reply:
fyi I remember doing something in that direction a while ago, but then had some issues and didn't have the time to finishh https://github.com/adobe/aio-lib-web/pull/104

Reply:
Filed an issue https://github.com/adobe/aio-cli-plugin-app/issues/364 and resolving this thread